File: Very_basic_examples/switch-case2.c
// Assembly code: https://c.godbolt.org/z/vzn8afG1E
#include <stddef.h>
#include <stdint.h>

#define PROGLEN 0x10000

enum OpCode {
    Add133,
    Sub133,
    Add155,
    Sub155,
    Add177,
    Sub177,
    RegToPc,
    Ret,
};

int32_t interpret_example(enum OpCode opcodes[PROGLEN]) {
    size_t instrptr=0;
    int32_t reg=0;

    while (1) {
        instrptr %= PROGLEN;
        switch (opcodes[instrptr++] & 0x07) {     // & 0x07 - faster!
            case Add133: reg += 133; break;
            case Sub133: reg -= 133; break;
            case Add155: reg += 155; break;
            case Sub155: reg -= 155; break;
            case Add177: reg += 177; break;
            case Sub177: reg -= 177; break;
            case RegToPc: instrptr = reg; break; // JMP
            case Ret: return reg; break;
        }
    }
}

File: Very_basic_examples/switch-case2.rs
// Assembly code: https://rust.godbolt.org/z/M9Pj3Kvja
const PROGLEN: usize = 0x10_000;

enum OpCode {
    Add133,
    Sub133,
    Add155,
    Sub155,
    Add177,
    Sub177,
    RegToPc,
    Ret,
}

fn interpret_example(opcodes: &[OpCode; PROGLEN]) -> i32 {
    let mut instrptr = 0;
    let mut reg: i32 = 0;

    loop {
        instrptr %= PROGLEN;
        match opcodes[instrptr] {
            OpCode::Add133 => reg += 133,
            OpCode::Sub133 => reg -= 133,
            OpCode::Add155 => reg += 155,
            OpCode::Sub155 => reg -= 155,
            OpCode::Add177 => reg += 177,
            OpCode::Sub177 => reg -= 177,
            OpCode::RegToPc => instrptr = reg as usize, // JMP
            OpCode::Ret => return reg,
        }
        instrptr += 1;
    }
}
